From 8670cc0997b376a3afcfdc157fbb13d1d81f7be8 Mon Sep 17 00:00:00 2001
From: Bernhard Mallinger <bernhard.mallinger@eox.at>
Date: Wed, 12 Jul 2023 10:45:55 +0200
Subject: [PATCH] Allow specifying `job_id` via request parameter

This is somewhat unconventional with REST design and leaves it up to the
process implementation to deal with duplicates and invalid IDs, however
it can make sense in certain use cases.

An alternative design would be to encode the desired `job_id` in the url:
`POST /processes/<process_id>/jobs/<job_id>`
This would suggest full control over the `job_id` from the client side,
so I'd prefer to pass the id via POST data. This way we can treat it as
suggestion, or possibly as a template from which the server derives the
actual id.
---
 pygeoapi/api.py                  | 3 ++-
 pygeoapi/process/manager/base.py | 5 +++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git pygeoapi/api.py pygeoapi/api.py
index e388275..b5449fa 100644
--- pygeoapi/api.py
+++ pygeoapi/api.py
@@ -3552,6 +3552,7 @@ class API:
 
         data_dict = data.get('inputs', {})
         LOGGER.debug(data_dict)
+        desired_job_id = data.get("job_id")
 
         try:
             execution_mode = RequestedProcessExecutionMode(
@@ -3562,7 +3563,7 @@ class API:
         try:
             LOGGER.debug('Executing process')
             result = self.manager.execute_process(
-                process_id, data_dict, execution_mode=execution_mode)
+                process_id, data_dict, execution_mode=execution_mode, desired_job_id=desired_job_id)
             job_id, mime_type, outputs, status, additional_headers = result
             headers.update(additional_headers or {})
             headers['Location'] = f'{self.base_url}/jobs/{job_id}'
diff --git pygeoapi/process/manager/base.py pygeoapi/process/manager/base.py
index d0d7824..0f6d35a 100644
--- pygeoapi/process/manager/base.py
+++ pygeoapi/process/manager/base.py
@@ -312,7 +312,8 @@ class BaseManager:
             self,
             process_id: str,
             data_dict: dict,
-            execution_mode: Optional[RequestedProcessExecutionMode] = None
+            execution_mode: Optional[RequestedProcessExecutionMode] = None,
+            desired_job_id: str | None = None
     ) -> Tuple[str, Any, JobStatus, Optional[Dict[str, str]]]:
         """
         Default process execution handler
@@ -329,7 +330,7 @@ class BaseManager:
                   response
         """
 
-        job_id = str(uuid.uuid1())
+        job_id = desired_job_id or str(uuid.uuid1())
         processor = self.get_processor(process_id)
         if execution_mode == RequestedProcessExecutionMode.respond_async:
             job_control_options = processor.metadata.get(
-- 
2.34.1

